src/logging/wal_recovery.cpp

```diff
@@ -0,0 +1,589 @@
+
+        // insert tuple
+        auto tuple_id = tg->InsertTupleFromRecovery(commit_id, tg_offset, tuple.get());
+        table->IncreaseTupleCount(1);
+        if (tuple_id == tg->GetAllocatedTupleCount() - 1) {
+          // add a new tile group if the last slot has been taken
+          if (table->GetTileGroupById(tg->GetTileGroupId() + 1).get() ==
+              nullptr)
+            table->AddTileGroupWithOidForRecovery(tg->GetTileGroupId() + 1);
+          catalog::Manager::GetInstance().GetNextTileGroupId();
+        }
+      }
+    }
+
+    else if(record_type==LogRecordType::TUPLE_DELETE) {
+      LOG_INFO("Tuple Delete");
+
+      oid_t database_id = (oid_t)record_decode.ReadLong();
+      oid_t schema_id = (oid_t) record_decode.ReadLong();
+      oid_t table_id = (oid_t)record_decode.ReadLong();
+
+      oid_t tg_block = (oid_t)record_decode.ReadLong();
+      oid_t tg_offset = (oid_t)record_decode.ReadLong();
+
+      auto table = storage::StorageManager::GetInstance()->GetTableWithOid(
+              database_id, table_id);
+      auto tg = table->GetTileGroupById(tg_block);
+
+      // This code might be useful on drop
+      if (schema_id == CATALOG_SCHEMA_OID) {  // catalog schema oid
+        switch (table_id) {
+
+          case DATABASE_CATALOG_OID:{ //pg_database
+            // TODO(graghura): delete database
+          }
+
+          case TABLE_CATALOG_OID: {  // pg_table
+            auto db_oid = tg->GetValue(tg_offset, 2).GetAs<oid_t>();
+            auto table_oid = tg->GetValue(tg_offset, 0).GetAs<oid_t>();
+            auto database =
+                    storage::StorageManager::GetInstance()->GetDatabaseWithOid(
+                            db_oid);  // Getting database oid from pg_table
+            database->DropTableWithOid(table_oid);
+            break;
+          }
+
+          case INDEX_CATALOG_OID: { // pg_index
+            // TODO(graghura): delete index
+          }
+
+          case COLUMN_CATALOG_OID: { // pg_column
+            // TODO(graghura): delete column (I think schema updates are not supported now)
+          }
+
+        }
+      }
+
+      auto tuple_id = tg->DeleteTupleFromRecovery(commit_id, tg_offset);
+      table->IncreaseTupleCount(1);
+      if (tuple_id == tg->GetAllocatedTupleCount() - 1) {
+        if (table->GetTileGroupById(tg->GetTileGroupId() + 1).get() ==
+            nullptr)
+          table->AddTileGroupWithOidForRecovery(tg->GetTileGroupId() + 1);
+        catalog::Manager::GetInstance().GetNextTileGroupId();
+      }
+      break;
+    }
+
+    else if(record_type==LogRecordType::TUPLE_UPDATE) {
+      LOG_INFO("Replaying TUPLE_UPDATE");
+    }
+
+    else if(record_type==LogRecordType::TRANSACTION_BEGIN) {
+      LOG_INFO("Replaying TXN_BEGIN");
+    }
+
+    else if(record_type==LogRecordType::TRANSACTION_COMMIT) {
+      LOG_INFO("Replaying TXN_COMMIT");
+    }
+
+    else if(record_type==LogRecordType::TRANSACTION_ABORT) {
```

Comment:
This check might need to be moved to ```ParseFromDisk``` and files an error there.
If a ```TRANSACTION_ABORT``` record were found at then at this time, then the previous changes in this transaction would have been made (which should not be, if I got this correctly).
On the other hand, if the transaction could reach here, then it should also have a ```TRANSACTION_COMMIT``` record.  I wonder if it is possible for a transaction to have both records. If not, then such check can be redundant.